### PR TITLE
Add heavy mix test for the scheduler

### PR DESCRIPTION
**Pull request by alexey-katranov**
The test found a couple of issues that are also addressed in the patch

### CODE REVIEW

File: src/tbb/governor.h [hunk 1]
```diff
@@ -53,9 +53,6 @@ class governor {
     //! TLS for scheduler instances associated with individual threads
     static basic_tls<thread_data*> theTLS;
 
-    //! Caches the maximal level of parallelism supported by the hardware
-    static unsigned DefaultNumberOfThreads;
-
     //! Caches the size of OS regular memory page
     static std::size_t DefaultPageSize;
```
Comment (phprus) [inline src/tbb/governor.h:57]: Unused variable.

File: test/tbb/test_task_arena.cpp [hunk 2]
```diff
@@ -2038,3 +2039,22 @@ TEST_CASE("is_inside_task in arena::execute") {
...
 #endif //__TBB_PREVIEW_TASK_GROUP_EXTENSIONS
+#endif
+
+//! The test for a regression when producer of affinity tasks and consumer left
+//! the arena and no one can proceed the tasks
+//! \brief \ref regression
+TEST_CASE("abandon afifinity tasks deadlock") {
+    const int num_arenas = 16;
+    std::vector<tbb::task_arena> arenas(num_arenas);
+    for (int trail = 0; trail < 1000000; trail) {
```
Comment (phprus) [inline test/tbb/test_task_arena.cpp:2050]: Maybe ```trail++```

File: test/tbb/test_task_arena.cpp [hunk 3]
```diff
@@ -1749,6 +1749,7 @@ struct enqueue_test_helper {
     std::atomic<std::size_t>& my_task_counter;
 };
 
+#if 0
```
Comment (vlserov) [inline test/tbb/test_task_arena.cpp:1752]: Do we need test below?

File: test/tbb/test_task_arena.cpp (hunk 2 above)
Comment (alexey-katranov) [reply test/tbb/test_task_arena.cpp:2050]: It was incorrect search and replace... Overall, the test does not reproduce the supposed issue and I will remove it completely.

File: test/tbb/test_task_arena.cpp (hunk 3 above)
Comment (alexey-katranov) [reply test/tbb/test_task_arena.cpp:1752]: We need them, it is some experiments, I will revert

File: src/tbb/governor.h (hunk 1 above)
Comment (alexey-katranov) [reply src/tbb/governor.h:57]: Good catch, thank you!

**Review by pavelkumbrasev — COMMENTED** [review]
Patch seems ok

File: test/tbb/test_scheduler_mix.cpp [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+        return Guard(this);
+    }
+
+    void signalShutdown() {
+        mContinueFlag.store(false, std::memory_order_seq_cst);
+    }
+
+    void waitCompletion() {
+        utils::SpinWaitUntilEq(mContinueFlag, false);
+    }
```
Comment (alexey-katranov) [inline test/tbb/test_scheduler_mix.cpp:368]: The logic seems broken.

File: src/tbb/arena_slot.cpp [hunk 5]
```diff
@@ -175,7 +175,9 @@ d1::task* arena_slot::steal_task(arena& a, isolation_type isolation) {
                 }
                 task_proxy& tp = *static_cast<task_proxy*>(result);
                 // If mailed task is likely to be grabbed by its destination thread, skip it.
-                if ( !(task_proxy::is_shared( tp.task_and_tag ) && tp.outbox->recipient_is_idle()) ) {
+
+                bool owner_worker_idle = a.mailbox(slot_index).recipient_is_idle();
+                if (!(task_proxy::is_shared(tp.task_and_tag) && tp.outbox->recipient_is_idle()) || owner_worker_idle) {
```
Comment (pavelkumbrasev) [inline src/tbb/arena_slot.cpp:180]:
```suggestion
                if (!task_proxy::is_shared(tp.task_and_tag) || !tp.outbox->recipient_is_idle()) || a.mailbox(slot_index).recipient_is_idle()) {
```

**Review by pavelkumbrasev — APPROVED** [review]

**Pull request merged**